### ISSUE TITLE
Garbled text

### ISSUE THREAD

**Issue by bytecode-nu** [rds1983/NuklearSharp#2]
I get, during my loading screen, garbled text. This only happens the very first loading screen, perhaps I am doing something silly. The very first frame(s) (of the first loading screen), the text is drawn correctly but it quickly becomes garbled. It looks like this: (It should print "This string is not drawn correctly", the string is too long but that's not the issue)

![First frame(s)](https://i.gyazo.com/f0e6614bae4a490d7162a3de4ca360d2.png)
![A new language](https://i.gyazo.com/60ecbcfaed774da5b6f16c3689d5e666.png)
![Boxes](https://i.gyazo.com/4a72c3e83dbfe49d6691454ce8bba888.png)
![Colourful](https://i.gyazo.com/aba56768ed992a8638a664b70ede6311.png)

(Each image represents a few frames, i.e. the text becomes destroyed over time/after a few frames.)

This only happens the very first loading screen. The main menu works properly, the progress bar on the loading screen is displayed correctly, and the in-game GUI is fine. For some reason, this only happens during the loading screen. 

A few notes:
* Latest Nuget version, 0.2.0.9. 
* I am constructing several NuklearContexts, one for each scene (main menu, loading screen, in-game).
* Scene transitions are done in the update, not in draw. That is, I ensure that every Begin has an End, and that a Draw follows. 

A few observations:
* Since I construct several contexts, the font is baked several times. Is this a problem?
* Notice that the label is correct the first few frames but gets garbled after a few frames. This makes me believe that the font is initialized properly but that it gets destroyed, somehow. 
* The shape of each colourful box kinda corresponds to the shape of the given letter. Not sure what this means. 
* Again, only the first loading screen has garbled text. If I go back to the menu scene from the in-game scene and start a new game, then the text is drawn correctly in the loading screen. 

It's a bit difficult to present a minimal example. I'll post some code below to demonstrate the flow of my game: (I hope I have included enough code.)

<details>

The very start of my update and draw loop, it simply calls the update or draw function of the scene manager which delegates the update or draw to the current scene. If there is an UI, then we make sure to call NuklearContext.Draw last. 

```cs
        protected override void Update(GameTime gameTime)
        {
            if (GamePad.GetState(PlayerIndex.One).Buttons.Back == ButtonState.Pressed || Keyboard.GetState().IsKeyDown(Keys.Escape))
                Exit();

            if (_sceneManager.IsEmpty) Exit();
            else _sceneManager.Update(gameTime); // The current scene might change after calling this. 
            base.Update(gameTime);
        }

        protected override void Draw(GameTime gameTime)
        {
            // This is not needed since the "ActiveUI", which is a NuklearContext, never changes during the Draw-call. 
            var currentUI = ActiveUI; 
            _sceneManager.Draw(gameTime);
            if (DrawUI) currentUI?.Draw();

            base.Draw(gameTime);
        }

```

The code of some functions of my main menu. Not very exciting, the important detail is that we only change scenes in Update(). 

```cs

        // Initialize the UI
        public override void Load()
        {
            UI = new NuklearContext(GraphicsDevice);
            var fontAtlas = new FontAtlasWrapper(UI);
            _header = fontAtlas.AddDefaultFont(42);
            _body = fontAtlas.AddDefaultFont(22);
            fontAtlas.Bake();
            UI.SetFont(_body);
        }

        // If the user has pressed a button in the UI, then we change the scene.
        public override void Update(GameTime gameTime)
        {
            switch (_choice)
            {
                case MenuChoice.NewGame:
                    GameSceneManager.LoadAndPush(
                        new LoadingScene<BoardScene>(
                            GraphicsDevice,
                            new BoardScene(Game, Content, GraphicsDevice, GameSceneManager),
                            GameSceneManager),
                        autoUnload: true, update: gameTime);
                    break;

                case MenuChoice.ExitGame:
                    GameSceneManager.Pop(GameSceneType.Exclusive);
                    break;
            }

            _choice = MenuChoice.Nothing;
        }

        // Draw the UI and if the user has pressed a button, set _choice to remember it. 
        public override void Draw(GameTime gameTime)
        {
            GraphicsDevice.Clear(Color.Black);

            var presentation = GraphicsDevice.PresentationParameters;
            var height = presentation.BackBufferHeight;

            if (UI.BeginTitled("menu", "Menu", presentation.Bounds, 0))
            {
                UI.LayoutRowDynamic(height / 4f, 1);
                UI.LayoutRowDynamic(30, 1);
                UI.SetFont(_header);
                UI.Text("Game Name", Nuklear.NK_TEXT_CENTERED);
                UI.SetFont(_body);
                UI.LayoutRowDynamic(30, 3);
                UI.Spacing(1);
                if (UI.ButtonText("New Game"))
                {
                    _choice = MenuChoice.NewGame;
                }
                UI.Spacing(2);
                if (UI.ButtonText("Exit"))
                {
                    _choice = MenuChoice.ExitGame;
                }
                UI.Spacing(1);
            }
            UI.End();
        }

```

The code of my simple loading screen. It loads the next scene asynchronously on another thread. Again, we change scene in Update() (but that does not matter since we have garbled text here). 

```cs

using System;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using NuklearSharp;
using NuklearSharp.MonoGame;

namespace Tower_Defence.GameScenes.Scenes
{
    public class LoadingScene<T> : GameScene, IDisposable
        where T : GameScene, ILoadableScene
    {
        private GraphicsDevice GraphicsDevice { get; set; }
        private GameSceneManager GameSceneManager { get; set; }
        private NuklearContext UI { get; set; }
        private bool AsyncLoading { get; }

        private Task _sceneToLoadTask;
        private readonly T _sceneToLoad;

        public LoadingScene(GraphicsDevice graphicsDevice, T sceneToLoad, GameSceneManager gameSceneManager, bool asyncLoading = true)
        {
            _sceneToLoad = sceneToLoad;
            GraphicsDevice = graphicsDevice;
            GameSceneManager = gameSceneManager;
            AsyncLoading = asyncLoading;
        }

        public override void Load()
        {
            UI = new NuklearContext(GraphicsDevice);
            var fontAtlas = new FontAtlasWrapper(UI);
            var font = fontAtlas.AddDefaultFont(26);
            fontAtlas.Bake();
            UI.SetFont(font);

            if (AsyncLoading)
            {
                _sceneToLoadTask = new Task(() => _sceneToLoad.Load());
                _sceneToLoadTask.Start();
            }
            else
            {
                _sceneToLoad.Load();
            }
        }

        public override void Unload()
        {
            this.Dispose();

            GraphicsDevice = null;
            GameSceneManager = null;
        }

        // Set this UI as the active UI. 
        public override void Resume()
        {
            TowerDefence.ActiveUI = UI;
        }

        // Pause is always called before Resume. So if the current scene has an UI, then it will set it. 
        public override void Pause()
        {
            TowerDefence.ActiveUI = null;
        }

        // Update doesn't really do much, it just checks if the asynchronous task has finished loading. "Busy waiting".
        public override void Update(GameTime gameTime)
        {
            // Check if we are done loading
            if (!AsyncLoading || _sceneToLoadTask.IsCompleted)
            {
                Debug.WriteLine($"Loading Completed!");

                // Check if there was an error whilst loading
                if (AsyncLoading && _sceneToLoadTask.IsFaulted)
                {
                    Debug.WriteLine("Loading Faulted!\n");
                    Debug.WriteLine(_sceneToLoadTask.Exception);

                    foreach (var s in GameSceneManager.Pop(GameSceneType.Exclusive).Where(s => !s.Unloaded))
                    {
                        s.Scene.Unload();
                        if (s.Scene is IDisposable d)
                            d.Dispose();
                    }
                }
                else
                {
                    // GameScene loaded correctly. 
                    // The new scene will draw this frame, but hasn't updated: update!
                    var scenes = GameSceneManager.Switch(_sceneToLoad, autoUnload: true, update: gameTime);
                    foreach (var scene in scenes.Where(s => !s.Unloaded)) scene.Scene.Unload();
                }
            }
        }

        // Draws the UI, i.e., the constant string "Hello, World! This string is not drawn correctly. ". 
        public override void Draw(GameTime gameTime)
        {
            GraphicsDevice.Clear(Color.Black);

            var presentation = GraphicsDevice.PresentationParameters;
            var height = presentation.BackBufferHeight;
            var width = presentation.BackBufferWidth;

            var screen = new Rectangle
            {
                X = width / 3,
                Y = height / 3,
                Width = width / 3,
                Height = height / 3
            };

            if (UI.BeginTitled("loading", "Loading", screen, 0))
            {
                // The progress bar is drawn correctly, only text is messed up. 
                //UI.LayoutRowDynamic(height / 6f, 1);
                UI.LayoutRowDynamic(30, 1);
                //UI.Progress(ref _progress, 100, Nuklear.nk_false);
                //UI.LayoutRowDynamic(30, 1);
                UI.Label("Hello, World! This string is not drawn correctly. ", Nuklear.NK_TEXT_LEFT);
            }
            UI.End();
        }

        public void Dispose()
        {
            _sceneToLoadTask?.Dispose();
        }
    }
}

```

</details>

**Comment by rds1983** [comment]
Maybe NuklearContext is being drawn while being disposed in the another thread?

**Comment by bytecode-nu** [comment]
I suspect that this is a difficult one to figure out. I'll try to strip away all unnecessary code at some point in order to provide a minimal example. I'm not sure if the bug lies in my code or not. It might also be an error in the original source code (I've noticed a few issues that carried over to this port), but I haven't found an issue regarding garbled text. 

I have tried to reuse the same NuklearContext without baking a new font, it gives me the same result. In other words, normal text in the menu and garbled text in the loading screen, using the same NuklearContext. Strangely enough, it only happens in the very first loading screen, subsequent loading screens looks good. 

I'll work on stripping away my code to provide a minimal example, when I find the time :-)
